src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -310,33 +331,59 @@ internal override sealed bool TryReadAsObject(ref Utf8JsonReader reader, JsonSer
         }
 
         /// <summary>
-        /// Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Performance optimization. Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Although the JIT added support to detect this IL pattern (box+isinst+br), it still manifests in TryWrite().
+        /// </summary>
+        internal virtual bool IsNull(in T value) => value is null;
+
+        /// <summary>
+        /// Performance optimization to determine if IsNull() above should be called.
         /// </summary>
-        internal virtual bool IsNull(in T value) => value == null;
+        /// <returns></returns>
+        internal virtual bool IsNullableOfT() => false;
 
+#if NET6_0_OR_GREATER
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+#endif
         internal bool TryWrite(Utf8JsonWriter writer, in T value, JsonSerializerOptions options, ref WriteStack state)
         {
             if (writer.CurrentDepth >= options.EffectiveMaxDepth)
             {
                 ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.EffectiveMaxDepth);
             }
 
-            if (default(T) is null && !HandleNullOnWrite && IsNull(value))
+            // We do not pass null values to converters unless HandleNullOnWrite is true. Null values for properties were
+            // already handled in GetMemberAndWriteJson() so we don't need to check for IgnoreNullValues here.
+            if (_autoWriteNulls)
```

Comment:
Note that `default(T) is null` is a JIT constant so I don't believe `_autoWriteNulls` is needed. Changing to
```suggestion
            if (default(T) is null && !HandleNullOnWrite)
```
Has the combined effect of eliminating the branch for non-nullable structs and only reading one field.

Reply:
Thanks. See other comment below related to `in` to avoid boxing -- combining both your suggestions I think is the best approach.